Implement missing functionality of AudioBufferSourceNode

collares opened the pull request:
<!-- Please describe your changes on the following line: -->
This is a tiny PR to support [Media servo/media#293](https://github.com/servo/media/pull/293). Last time I ran WPT there was only one failing AudioBufferSourceNode test (which I will debug later this week). Once I rebased, however, I started getting unrelated servo-media build failures:

```
error[E0599]: no method named `shutdown_audio_context` found for type `std::sync::Arc<servo_media::ServoMedia>` in the current scope
   --> components/script/dom/baseaudiocontext.rs:566:14

error[E0599]: no method named `shutdown_player` found for type `std::sync::Arc<servo_media::ServoMedia>` in the current scope
    --> components/script/dom/htmlmediaelement.rs:1866:18
```

I wanted to get this PR up so https://github.com/servo/media/pull/293 can be tested. When the unrelated build failures are fixed, I will run `./mach update-wpt` and update this PR.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` only reports unrelated errors (broken servo/media master?)
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #22363

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23997)
<!-- Reviewable:end -->

collares committed:
AudioBuffer: Track sample rate

collares committed:
AudioBufferSourceNode: Support start parameters (offset, duration)

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/audiobuffer.rs, components/script/dom/audiobuffersourcenode.rs
 * @kichjang: components/script/dom/audiobuffer.rs, components/script/dom/audiobuffersourcenode.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

ferjm commented:
>  Once I rebased, however, I started getting unrelated servo-media build failures

This should be fixed by #24000

collares committed:
webaudio: update wpt test expectations

collares commented:
Thanks! Rebased on top of your patch, and updated test expectations for the whole webaudio directory (manually tidying up cases where the only change was the order of the failures). I can revert the unrelated WPT expectation changes if you prefer.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 65e48f061630f671959544778543317a2b76f061 with merge c4fedf118b1e77840ead8c92baf0250bc2e8c4a8...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13209)

collares commented:
I didn't know if updating Cargo.lock to use my branch was the right thing to do, so I didn't do it. Sorry! Added a commit with the correct overrides, so bors should now be able to run WPT.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24000) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24002) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
r? @ferjm

ferjm reviewed (approved):
Great work! r=me updating to the proper servo-media revision. Thanks!

collares committed:
AudioBufferSourceNode: Send loop messages

collares committed:
webaudio: update servo-media revision

collares commented:
Updated. Many thanks for the review, and for being understanding about the delay. I think taking the time to read the spec carefully paid off, as I got to file three (all of them minor, hopefully not bogus) spec clarification issues :)

jdm commented:
@bors-servo r=ferjm

bors-servo commented:
:pushpin: Commit 8ee2b7d has been approved by `ferjm`

<!-- @bors-servo r=ferjm 8ee2b7d8705501b3ef14f7694b1ec269ecd3e170 -->

bors-servo commented:
:hourglass: Testing commit 8ee2b7d8705501b3ef14f7694b1ec269ecd3e170 with merge 5cbef0fa788c2af4ba6fa7227d758d04877322d1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13162)

CYBAI commented:
<details>

```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: expected true got false", 
    "stack": "_logFailed/<@http://web-platform.test:8000/webaudio/resources/audit.js:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n_logFailed@http://web-platform.test:8000/webaudio/resources/audit.js:49:5\n_finalize@http://web-platform.test:8000/webaudio/resources/audit.js:215:9\n_assert@http://web-platform.test:8000/webaudio/resources/audit.js:228:12\nbeCloseToArray@http://web-platform.test:8000/webaudio/resources/audit.js:1019:19\nrunTest/<@http://web-platform.test:8000/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html:258:18\n", 
    "subtest": "X 1-channel [0, 0, 1] -> [0, 0, 10000]: distanceModel: inverse, rolloff: 1, left channel does not equal [0.7071067690849304,0.7071067690849304,0.013253915123641491,0.00668965233489871,0.004473877139389515,0.003360723378136754,0.0026911369059234858,0.0022440373431891203,0.0019243330461904407,0.0016843645134940743,0.0014976091915741563,0.001348133897408843,0.0012257886119186878,0.001123802037909627,0.0010374825214967132,0.0009634775342419744...] with an element-wise tolerance of {\"absoluteThreshold\":0,\"relativeThreshold\":0}.\n\tIndex\tActual\t\t\tExpected\t\tAbsError\t\tRelError\t\tTest threshold\n\t[0]\t0.0000000000000000e+0\t7.0710676908493042e-1\t7.0710676908493042e-1\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[1]\t0.0000000000000000e+0\t7.0710676908493042e-1\t7.0710676908493042e-1\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[2]\t0.0000000000000000e+0\t1.3253915123641491e-2\t1.3253915123641491e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[3]\t0.0000000000000000e+0\t6.6896523348987103e-3\t6.6896523348987103e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[4]\t0.0000000000000000e+0\t4.4738771393895149e-3\t4.4738771393895149e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t...and 251 more errors.\n\tMax AbsError of 7.0710676908493042e-1 at index of 0.\n\tMax RelError of 1.0000000000000000e+0 at index of 0.\n", 
    "test": "/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html", 
    "line": 253553, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: expected true got false", 
    "stack": "_logFailed/<@http://web-platform.test:8000/webaudio/resources/audit.js:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n_logFailed@http://web-platform.test:8000/webaudio/resources/audit.js:49:5\n_finalize@http://web-platform.test:8000/webaudio/resources/audit.js:215:9\n_assert@http://web-platform.test:8000/webaudio/resources/audit.js:228:12\nbeCloseToArray@http://web-platform.test:8000/webaudio/resources/audit.js:1019:19\nrunTest/<@http://web-platform.test:8000/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html:260:16\n", 
    "subtest": "X 1-channel [0, 0, 1] -> [0, 0, 10000]: distanceModel: inverse, rolloff: 1, right channel does not equal [0.7071067690849304,0.7071067690849304,0.013253915123641491,0.00668965233489871,0.004473877139389515,0.003360723378136754,0.0026911369059234858,0.0022440373431891203,0.0019243330461904407,0.0016843645134940743,0.0014976091915741563,0.001348133897408843,0.0012257886119186878,0.001123802037909627,0.0010374825214967132,0.0009634775342419744...] with an element-wise tolerance of {\"absoluteThreshold\":0,\"relativeThreshold\":0}.\n\tIndex\tActual\t\t\tExpected\t\tAbsError\t\tRelError\t\tTest threshold\n\t[0]\t0.0000000000000000e+0\t7.0710676908493042e-1\t7.0710676908493042e-1\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[1]\t0.0000000000000000e+0\t7.0710676908493042e-1\t7.0710676908493042e-1\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[2]\t0.0000000000000000e+0\t1.3253915123641491e-2\t1.3253915123641491e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[3]\t0.0000000000000000e+0\t6.6896523348987103e-3\t6.6896523348987103e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[4]\t0.0000000000000000e+0\t4.4738771393895149e-3\t4.4738771393895149e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t...and 251 more errors.\n\tMax AbsError of 7.0710676908493042e-1 at index of 0.\n\tMax RelError of 1.0000000000000000e+0 at index of 0.\n", 
    "test": "/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html", 
    "line": 253554, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: expected true got false", 
    "stack": "_logFailed/<@http://web-platform.test:8000/webaudio/resources/audit.js:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n_logFailed@http://web-platform.test:8000/webaudio/resources/audit.js:49:5\n_finalize@http://web-platform.test:8000/webaudio/resources/audit.js:215:9\n_assert@http://web-platform.test:8000/webaudio/resources/audit.js:228:12\nbeCloseToArray@http://web-platform.test:8000/webaudio/resources/audit.js:1019:19\nrunTest/<@http://web-platform.test:8000/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html:258:18\n", 
    "subtest": "X 2-channel [0, 0, 1] -> [0, 0, 10000]: distanceModel: inverse, rolloff: 1, left channel does not equal [1,1,0.018743867054581642,0.009460597299039364,0.006327017676085234,0.0047527807764709,0.0038058424834161997,0.003173548262566328,0.0027214179281145334,0.0023820511996746063,0.002117939293384552,0.0019065493252128363,0.0017335269367322326,0.001589296036399901,0.0014672218821942806,0.0013625629944726825...] with an element-wise tolerance of {\"absoluteThreshold\":0,\"relativeThreshold\":0}.\n\tIndex\tActual\t\t\tExpected\t\tAbsError\t\tRelError\t\tTest threshold\n\t[0]\t0.0000000000000000e+0\t1.0000000000000000e+0\t1.0000000000000000e+0\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[1]\t0.0000000000000000e+0\t1.0000000000000000e+0\t1.0000000000000000e+0\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[2]\t0.0000000000000000e+0\t1.8743867054581642e-2\t1.8743867054581642e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[3]\t0.0000000000000000e+0\t9.4605972990393639e-3\t9.4605972990393639e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[4]\t0.0000000000000000e+0\t6.3270176760852337e-3\t6.3270176760852337e-3\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t...and 251 more errors.\n\tMax AbsError of 1.0000000000000000e+0 at index of 0.\n\tMax RelError of 1.0000000000000000e+0 at index of 0.\n", 
    "test": "/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html", 
    "line": 253557, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: expected true got false", 
    "stack": "_logFailed/<@http://web-platform.test:8000/webaudio/resources/audit.js:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n_logFailed@http://web-platform.test:8000/webaudio/resources/audit.js:49:5\n_finalize@http://web-platform.test:8000/webaudio/resources/audit.js:215:9\n_assert@http://web-platform.test:8000/webaudio/resources/audit.js:228:12\nbeCloseToArray@http://web-platform.test:8000/webaudio/resources/audit.js:1019:19\nrunTest/<@http://web-platform.test:8000/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html:260:16\n", 
    "subtest": "X 2-channel [0, 0, 1] -> [0, 0, 10000]: distanceModel: inverse, rolloff: 1, right channel does not equal [2,2,0.037487734109163284,0.018921194598078728,0.012654035352170467,0.0095055615529418,0.007611684966832399,0.006347096525132656,0.005442835856229067,0.004764102399349213,0.004235878586769104,0.0038130986504256725,0.003467053873464465,0.003178592072799802,0.0029344437643885612,0.002725125988945365...] with an element-wise tolerance of {\"absoluteThreshold\":0,\"relativeThreshold\":0}.\n\tIndex\tActual\t\t\tExpected\t\tAbsError\t\tRelError\t\tTest threshold\n\t[0]\t0.0000000000000000e+0\t2.0000000000000000e+0\t2.0000000000000000e+0\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[1]\t0.0000000000000000e+0\t2.0000000000000000e+0\t2.0000000000000000e+0\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[2]\t0.0000000000000000e+0\t3.7487734109163284e-2\t3.7487734109163284e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[3]\t0.0000000000000000e+0\t1.8921194598078728e-2\t1.8921194598078728e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t[4]\t0.0000000000000000e+0\t1.2654035352170467e-2\t1.2654035352170467e-2\t1.0000000000000000e+0\t0.0000000000000000e+0\n\t...and 251 more errors.\n\tMax AbsError of 2.0000000000000000e+0 at index of 0.\n\tMax RelError of 1.0000000000000000e+0 at index of 0.\n", 
    "test": "/webaudio/the-audio-api/the-pannernode-interface/panner-automation-position.html", 
    "line": 253558, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: expected true got false", 
    "stack": "_logFailed/<@http://web-platform.test:8000/webaudio/resources/audit.js:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n_logFailed@http://web-platform.test:8000/webaudio/resources/audit.js:49:5\n_finalize@http://web-platform.test:8000/webaudio/resources/audit.js:215:9\n_assert@http://web-platform.test:8000/webaudio/resources/audit.js:228:12\nbeCloseToArray@http://web-platform.test:8000/webaudio/resources/audit.js:1019:19\n@http://web-platform.test:8000/webaudio/the-audio-api/the-audioparam-interface/audioparam-method-chaining.html:134:20\n", 
    "subtest": "X The rendered envelope does not equal [0,0.000125,0.00025,0.000375,0.0005,0.000625,0.00075,0.000875,0.001,0.001125,0.00125,0.001375,0.0015,0.001625,0.00175,0.001875...] with an element-wise tolerance of {\"absoluteThreshold\":0.0000040532,\"relativeThreshold\":0}.\n\tIndex\tActual\t\t\tExpected\t\tAbsError\t\tRelError\t\tTest threshold\n\t[1]\t0.0000000000000000e+0\t1.2500000000000000e-4\t1.2500000000000000e-4\t1.0000000000000000e+0\t4.0531999999999996e-6\n\t[2]\t1.2501562014222145e-4\t2.5000000000000001e-4\t1.2498437985777855e-4\t4.9993751943111420e-1\t4.0531999999999996e-6\n\t[3]\t2.5003124028444290e-4\t3.7500000000000001e-4\t1.2496875971555711e-4\t3.3325002590815228e-1\t4.0531999999999996e-6\n\t[4]\t3.7504688953049481e-4\t5.0000000000000001e-4\t1.2495311046950520e-4\t2.4990622093901041e-1\t4.0531999999999996e-6\n\t[5]\t5.0006248056888580e-4\t6.2500000000000001e-4\t1.2493751943111421e-4\t1.9990003108978274e-1\t4.0531999999999996e-6\n\t...and 31734 more errors.\n\tMax AbsError of 9.8985821103293703e-1 at index of 31999.\n\t[31999]\t1.0000000000000000e+0\t1.0141788967063026e-2\t9.8985821103293703e-1\t9.7601933371681213e+1\t4.0531999999999996e-6\n\tMax RelError of 9.7601933371681213e+1 at index of 31999.\n\t[31999]\t1.0000000000000000e+0\t1.0141788967063026e-2\t9.8985821103293703e-1\t9.7601933371681213e+1\t4.0531999999999996e-6\n", 
    "test": "/webaudio/the-audio-api/the-audioparam-interface/audioparam-method-chaining.html", 
    "line": 268502, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Test setValueAtTime with start time not on a block boundary", 
    "test": "/webaudio/the-audio-api/the-audioparam-interface/setValueAtTime-within-block.html", 
    "line": 306309, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Test setTargetAtTime after an event in the same processing block", 
    "test": "/webaudio/the-audio-api/the-audioparam-interface/setTargetAtTime-after-event-within-block.html", 
    "line": 339836, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

</details>

collares commented:
1) setValueAtTime-within-block.html and setTargetAtTime-after-event-within-block.html use looping AudioBufferSourceNodes as part of the test, so it makes sense that they started passing. I forgot about loop messages in my first patch, so I missed them in my WPT update.

2) panner-automation-position.html was previously failing and still fails, I just needed to update the error messages. It changed because it uses a looping AudioBufferSourceNodes for constructing the reference output. 

3) audioparam-method-chaining.html.ini was previously failing and still fails, I just needed to update the error messages. I am not sure why the output changed, but it is probably related to the fact that Tick::from_time now rounds up to the next Tick (previously it rounded to the nearest one), which is what makes sense for computing start_at and stop_at. There seems to be an off-by-one error in AudioParam somewhere, judging by the test output, but since no time values in the test are subsample-accurate I don't think this is caused by my change.

Given the above, I don't think any code changes are needed. I have updated the WPT results accordingly.

collares commented:
Filed https://github.com/servo/media/blob/eafa387be6424c99efc11231c9c5a03284116ddb/audio/param.rs#L149 for the off-by-one error in AudioParam.

ferjm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4d43f34 has been approved by `ferjm`

<!-- @bors-servo r=ferjm 4d43f34bac1e4c5d56f1339166186d8078754874 -->

bors-servo commented:
:hourglass: Testing commit 4d43f34bac1e4c5d56f1339166186d8078754874 with merge 5832781070a4b21925c1a9dfb55f93e86df9de50...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/eUgL3hLJQ8OQ1HWrpKAKtg)

collares commented:
There were some WebGL test failures:
<details>
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Overall test", 
    "test": "/_webgl/conformance/textures/misc/origin-clean-conformance-offscreencanvas.html", 
    "line": 31010, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/_webgl/conformance/textures/misc/origin-clean-conformance-offscreencanvas.html", 
    "line": 31011, 
    "action": "test_result", 
    "expected": "ERROR"
}
</details>

CYBAI commented:
@bors-servo retry

- #24038

bors-servo commented:
:hourglass: Testing commit 4d43f34bac1e4c5d56f1339166186d8078754874 with merge cacd4ed4e216ec4c572ea78c7e86bcd0b3d8c86b...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13184), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13245), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Kjq6lXu2TieVmSF9DpxrIQ)
Approved by: ferjm
Pushing cacd4ed4e216ec4c572ea78c7e86bcd0b3d8c86b to master...
